File: app/src/main/java/se/jakob/knarkklocka/data/AlarmListAdapter.java
package se.jakob.knarkklocka.data;

import android.content.Context;
import android.content.res.Resources;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.List;
import java.util.Locale;

import se.jakob.knarkklocka.R;

public class AlarmListAdapter extends RecyclerView.Adapter<AlarmListAdapter.AlarmViewHolder> {

    private final LayoutInflater mInflater;
    private List<Alarm> mAlarms; // Cached copy of alarms
    private Resources res;

    public AlarmListAdapter(Context context) {
        mInflater = LayoutInflater.from(context);
        res = context.getResources();
    }

    @NonNull
    @Override
    public AlarmViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = mInflater.inflate(R.layout.recyclerview_item, parent, false);
        return new AlarmViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull AlarmViewHolder holder, int position) {
        if (mAlarms != null) {
            Alarm current = mAlarms.get(position);
            SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm", Locale.getDefault());
            String dateString = dateFormat.format(current.getEndTime());
            String string = current.getId() + " " + dateString;
            holder.dueTimeItemView.setText(string);

            int state = current.getState();
            String stateString = current.getStateString(state);
            holder.stateItemView.setText(stateString);


            int snoozes = current.getSnoozes();
            String snoozeString = res.getQuantityString(R.plurals.snoozes, snoozes, snoozes);
            //String snoozeString = "Snoozes: " + snoozes;
            holder.snoozeItemView.setText(snoozeString);
        } else {
            // Covers the case of data not being ready yet.
            holder.dueTimeItemView.setText("No Alarms");
        }
    }

    public void setAlarms(List<Alarm> words) {
        mAlarms = words;
        notifyDataSetChanged();
    }

    // getItemCount() is called many times, and when it is first called,
    // mAlarms has not been updated (means initially, it's null, and we can't return null).
    @Override
    public int getItemCount() {
        if (mAlarms != null)
            return mAlarms.size();
        else return 0;
    }

    class AlarmViewHolder extends RecyclerView.ViewHolder {
        private final TextView dueTimeItemView;
        private final TextView stateItemView;
        private final TextView snoozeItemView;

        private AlarmViewHolder(View itemView) {
            super(itemView);
            dueTimeItemView = itemView.findViewById(R.id.tv_alarm_due);
            stateItemView = itemView.findViewById(R.id.tv_alarm_state);
            snoozeItemView = itemView.findViewById(R.id.tv_alarm_snoozes);
        }
    }
}

File: app/src/main/java/se/jakob/knarkklocka/data/AlarmListAdapter.kt
package se.jakob.knarkklocka.data
import android.content.Context
import android.content.res.Resources
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import se.jakob.knarkklocka.R
import java.text.SimpleDateFormat
import java.util.*

class AlarmListAdapter(context: Context) : RecyclerView.Adapter<AlarmListAdapter.AlarmViewHolder>() {

    private val mInflater: LayoutInflater = LayoutInflater.from(context)
    var mAlarms: List<Alarm>? = null // Cached copy of alarms
    set(alarms) {
        field = alarms
        notifyDataSetChanged()
    }
    private val res: Resources = context.resources

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AlarmViewHolder {
        val itemView = mInflater.inflate(R.layout.recyclerview_item, parent, false)
        return AlarmViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: AlarmViewHolder, position: Int) {
        mAlarms?.run {
            val currentAlarm = this[position]
            val dateString = SimpleDateFormat("HH:mm", Locale.getDefault()).format(currentAlarm.endTime)
            //val dateString = dateFormat.format(currentLiveAlarm.endTime)
            //val string = currentLiveAlarm.id.toString() + " " + dateString
            holder.dueTimeItemView.text = String.format(Locale.getDefault(), "%d: \t %s", currentAlarm.id, dateString)

            val stateString = currentAlarm.stateToString
            holder.stateItemView.text = stateString

            val snoozes = currentAlarm.snoozes
            val snoozeString = res.getQuantityString(R.plurals.snoozes, snoozes, snoozes)
            //String snoozeString = "Snoozes: " + snoozes;
            holder.snoozeItemView.text = snoozeString
        } ?: run {
            // Covers the case of data not being ready yet.
            holder.dueTimeItemView.text = res.getString(R.string.viewholder_no_alarms)
        }
    }

    // getItemCount() is called many times, and when it is first called,
    // mAlarms has not been updated (means initially, it's null, and we can't return null).
    override fun getItemCount(): Int {
        return mAlarms?.size ?: 0
    }

    inner class AlarmViewHolder constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val dueTimeItemView: TextView = itemView.findViewById(R.id.tv_alarm_due)
        val stateItemView: TextView = itemView.findViewById(R.id.tv_alarm_state)
        val snoozeItemView: TextView = itemView.findViewById(R.id.tv_alarm_snoozes)
    }
}